### ISSUE TITLE
About the deepspeed

### ISSUE THREAD

**Issue by LetheSec** [locuslab/tofu#36]
Hi,

Have you tried experimenting without using Deepspeed?  

When I delete the code and configuration related to Deepspeed in the repo, the results obtained will be quite different. 

The effective batch size is still set to 32 (i.e., batch_size=8 and gradient_accumulation_steps=4) and the GPUs I use are two A6000 48G. I directly use the command `CUDA_VISIBLE_DEVICES=0,1 python forget.py xxx`to run the script instead of `torchrun`.

I'm not sure the reason behind it. 

Any help is appreciated!

**Comment by LetheSec** [comment]
This seems to be a bug. 

When I  discard Deepspeed, the `add_special_tokens=True`  in the following line doesn't seem to work, that is, it actually does not add the `<s>` at the beginning of the sequence.
https://github.com/locuslab/tofu/blob/80159d8ea39edf147fb09cd82aefa08e506e6718/data_module.py#L13

It seems to be a more general approach using `tokenizer.encode` instead:

```python
num_question_tokens = len(tokenizer.encode(new_question, add_special_tokens=True))
```

Then, close results can be obtained by properly adjusting the learning rate.

**Comment by zaocan666** [comment]
Me too. When running without deepspeed, the training results in insufficient unleanring. The loss only declines slightly. Have you already solved the problem?

**Comment by zaocan666** [comment]
Me too. When running without deepspeed, the training results in insufficient unleanring. The loss only declines slightly. Have you already solved the problem?

**Comment by LetheSec** [comment]
> Me too. When running without deepspeed, the training results in insufficient unleanring. The loss only declines slightly. Have you already solved the problem?

I didn't find a solution and ended up following the original code in the repo. Without using deepspeed, it seems that one can only re-tune the hyper-parameters to achieve similar results, such as increasing the learning rate.

I speculated that the reason might be from the transformers library (not sure), refer to the link below:
https://mp.weixin.qq.com/s/hmIvSsZ2-yKEeiXUNRlTQw

**Comment by zaocan666** [comment]
> > Me too. When running without deepspeed, the training results in insufficient unleanring. The loss only declines slightly. Have you already solved the problem?
> 
> I didn't find a solution and ended up following the original code in the repo. Without using deepspeed, it seems that one can only re-tune the hyper-parameters to achieve similar results, such as increasing the learning rate.
> 
> I speculated that the reason might be from the transformers library (not sure), refer to the link below: https://mp.weixin.qq.com/s/hmIvSsZ2-yKEeiXUNRlTQw

Thank you for sharing. Whether using deepspeed makes a big difference on training results, which is strange.
